### PR TITLE
cli: allow to specify wallet via configuration file

### LINKED ISSUE #2487: wallet init without password
please add support to generate wallets with empty password, without asking it
```
neo-go wallet init -a -w wallet.json
```

### LINKED ISSUE #2536: support wallet password from config
As an automation engineer, I need to use `neo-go` cli in scripts and autotests. Wallets require to enter the password interactively. For this purpose I use `expect` of `pyexpect` python library. The issue is that interactive password entry needs to be adjusted for every environment and every OS -- at the moment it is Jenkins, bash and python script on Linux and MacOS.

If wallet password might be obtained from config non-interactively it will be a convenient feature for using `neo-go` in automation.

The proposal is to have the same way to configure defualt wallet, address and password for it, as we have in `neofs-cli` and corresponding tools

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2487, close #2536.

### CODE REVIEW

File: cli/wallet/wallet.go [hunk 1]
```diff
@@ -400,22 +426,24 @@ func convertWallet(ctx *cli.Context) error {
...
 	if err != nil {
 		return cli.NewExitError(err, 1)
 	}
 
 	defer wall.Close()
 
-	if err := createAccount(wall); err != nil {
+	// Do not use the provided password from wallet config file, otherwise it can be misleading
+	// (wallet may already have an account with the specified password).
+	if err := createAccount(wall, nil); err != nil {
```
Comment (fyrchik) [inline cli/wallet/wallet.go:438]: I think with `wallet-config` option this is not misleading at all. Actually, it is a feature, e.g. `change-password` command assumes all accounts have the same password.

File: cli/smartcontract/smart_contract.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/nspcc-dev/neo-go/cli/input"
 	"github.com/nspcc-dev/neo-go/cli/options"
 	"github.com/nspcc-dev/neo-go/cli/paramcontext"
+	cliwallet "github.com/nspcc-dev/neo-go/cli/wallet"
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:17]: If we're creating this dependency...

File: cli/smartcontract/smart_contract.go [hunk 3]
```diff
@@ -46,6 +47,10 @@ var (
 		Name:  "wallet, w",
 		Usage: "wallet to use to get the key for transaction signing",
 	}
+	walletConfigFlag = cli.StringFlag{
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:50]: ...then we can reuse the flag from the `cli/wallet` package...

File: cli/smartcontract/smart_contract.go [hunk 4]
```diff
@@ -818,9 +825,19 @@ func getAccFromContext(ctx *cli.Context) (*wallet.Account, *wallet.Wallet, error
 	var addr util.Uint160
 
 	wPath := ctx.String("wallet")
-	if len(wPath) == 0 {
+	walletConfigPath := ctx.String("wallet-config")
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:828]: ...and generalize `getAccFromContext` that will be reused by both packages.

File: cli/wallet/multisig.go [hunk 5]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 func signStoredTransaction(ctx *cli.Context) error {
-	wall, err := readWallet(ctx.String("wallet"))
+	wall, pass, err := readWallet(ctx.String("wallet"), ctx.String("wallet-config"))
```
Comment (roman-khimov) [inline cli/wallet/multisig.go:15]: Pass context to `readWallet`?

File: cli/wallet/wallet.go [hunk 6]
```diff
@@ -314,7 +333,7 @@ func claimGas(ctx *cli.Context) error {
 }
 
 func changePassword(ctx *cli.Context) error {
-	wall, err := openWallet(ctx.String("wallet"))
+	wall, oldPass, err := openWallet(ctx.String("wallet"), ctx.String("wallet-config"))
```
Comment (roman-khimov) [inline cli/wallet/wallet.go:336]: Does this option make any sense for `changePassword`? `wallet-config` is supposed to be used in non-interactive scenarios and here we're requesting a new password anyway (but we can't take it from the configuration either).

File: cli/wallet/wallet.go (hunk 1 above)
Comment (roman-khimov) [reply cli/wallet/wallet.go:438]: Yeah, reusing the password from the config is OK here.

Comment (AnnaShaleva) [reply cli/wallet/wallet.go]: Fixed.

File: cli/contract_test.go [hunk 7]
```diff
@@ -613,12 +614,19 @@ func TestComlileAndInvokeFunction(t *testing.T) {
...
+	}
+	yml, err := yaml.Marshal(cfg)
+	require.NoError(t, err)
+	require.NoError(t, os.WriteFile(configPath, yml, 0666))
 	e.Run(t, "neo-go", "contract", "deploy",
 		"--rpc-endpoint", "http://"+e.RPC.Addr, "--force",
-		"--wallet", validatorWallet, "--address", validatorAddr,
+		"--wallet-config", configPath, "--address", validatorAddr,
 		"--in", nefName, "--manifest", manifestName)
-
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/contract_test.go:621]: Fixed.

File: cli/smartcontract/smart_contract.go (hunk 3 above)
Comment (AnnaShaleva) [reply cli/smartcontract/smart_contract.go:50]: They have different usage texts, and this one is invocation-related. They also have different styles of usage text (capitalization). And we don't have any other exported flags, so it doesn't look goot to me. Maybe we can create a separate helper package and move `walletConfigFlag` and `walletFlag` there?

File: cli/smartcontract/smart_contract.go (hunk 4 above)
Comment (AnnaShaleva) [reply cli/smartcontract/smart_contract.go:828]: The case is that each wallet-related command performs different actions with account. Sometimes we need to unlock it, sometimes not. Sometimes we're OK with default account. Sometimes it's OK to have several accounts. Also, accounts are hidden under different flag names. And all of these cases must be handled inside this single method.

Maybe we can reuse the code that opens wallet, but I'd prefer to keep `getAccFromContext` inside `smartcontract` package and do not generalize it.

File: cli/wallet/multisig.go (hunk 5 above)
Comment (AnnaShaleva) [reply cli/wallet/multisig.go]: Fixed.

File: cli/wallet/wallet.go (hunk 6 above)
Comment (AnnaShaleva) [reply cli/wallet/wallet.go]: You're right, removed this flag from `changePassword`.

File: cli/wallet/wallet.go [hunk 8]
```diff
@@ -45,6 +47,10 @@ var (
 		Name:  "wallet, w",
 		Usage: "Target location of the wallet file ('-' to read from stdin).",
 	}
+	walletConfigFlag = cli.StringFlag{
+		Name:  "wallet-config",
+		Usage: "Target location of the wallet config file, overrides --wallet flag if specified.",
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply cli/wallet/wallet.go:52]: Yeah, having both of them doesn't make any sense, so we should return an error in this case.

[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/wallet/wallet.go:53]: Fixed.

File: cli/wallet/wallet.go [hunk 9]
```diff
@@ -35,16 +37,21 @@ const (
...
 )
 
 var (
 	walletPathFlag = cli.StringFlag{
 		Name:  "wallet, w",
 		Usage: "Target location of the wallet file ('-' to read from stdin).",
 	}
+	walletConfigFlag = cli.StringFlag{
+		Name:  "wallet-config",
+		Usage: "Target location of the wallet config file, overrides --wallet flag if specified.",
```
Comment (roman-khimov) [inline cli/wallet/wallet.go:53]: Not overrides.

File: cli/smartcontract/smart_contract.go [hunk 10]
```diff
@@ -33,19 +34,24 @@ import (
...
+	errNoSmartContractName    = errors.New("no name was provided, specify the '--name or -n' flag")
+	errFileExist              = errors.New("A file with given smart-contract name already exists")
 
 	walletFlag = cli.StringFlag{
 		Name:  "wallet, w",
 		Usage: "wallet to use to get the key for transaction signing",
 	}
+	walletConfigFlag = cli.StringFlag{
+		Name:  "wallet-config",
+		Usage: "path to wallet config to use to get the key for transaction signing; overrides --wallet flag if specified",
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/smart_contract.go]: Fixed, I've added a comment that it conflicts with `--wallet` flag.

File: cli/smartcontract/manifest.go [hunk 11]
```diff
@@ -16,10 +17,23 @@ import (
 
 func manifestAddGroup(ctx *cli.Context) error {
 	walletPath := ctx.String("wallet")
-	if len(walletPath) == 0 {
+	walletConfigPath := ctx.String("wallet-config")
+	if len(walletPath) != 0 && len(walletConfigPath) != 0 {
+		return errConflictingWalletFlags
+	}
+	if len(walletPath) == 0 && len(walletConfigPath) == 0 {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/manifest.go]: Yes, this code is moved to a single function completely.

File: cli/wallet/wallet.go (hunk 9 above)
Comment (AnnaShaleva) [reply cli/wallet/wallet.go]: Fixed.

File: cli/smartcontract/smart_contract.go [hunk 12]
```diff
@@ -387,12 +398,13 @@ func NewCommands() []cli.Command {
...
 							walletFlag,
+							walletConfigFlag,
 							cli.StringFlag{
 								Name:  "sender, s",
 								Usage: "deploy transaction sender",
 							},
-							cli.StringFlag{
-								Name:  "account, a",
+							flags.AddressFlag{
+								Name:  addressFlagName, // use the same name for handler code unification.
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:407]: Breaks backwards compatibility somewhat, but they're the same in short form and it's a very rarely used feature (and the only command that differs), so probably OK.

File: pkg/wallet/wallet.go [hunk 13]
```diff
@@ -168,9 +168,8 @@ func (w *Wallet) JSON() ([]byte, error) {
 	return json.MarshalIndent(w, " ", "	")
 }
 
-// Close closes the internal rw if its an io.ReadCloser.
-func (w *Wallet) Close() {
-}
+// Close performs nothing and is left for backwards compatibility.
```
Comment (roman-khimov) [inline pkg/wallet/wallet.go:171]: Mark as deprecated?

Comment (AnnaShaleva) [reply pkg/wallet/wallet.go:172]: Fixed.

Comment (AnnaShaleva) [reply pkg/wallet/wallet.go:172]: Can't we simply remove this method?

Comment (roman-khimov) [reply pkg/wallet/wallet.go:172]: Deprecate now, remove later. Not that it's very critical in this particular case (who calls it? removing a line is easy), but we better do it this way where we can.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**